### PR TITLE
GH-41899: [C++]  IPC: Minor enhance the code of writer

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Enhance the code of IPC writer

### What changes are included in this PR?

1. memcpy rather than reinterpret_cast
2. move the buffer rather than copying

### Are these changes tested?

Covered by existing

### Are there any user-facing changes?

No

* GitHub Issue: #41899

### CODE REVIEW

File: cpp/src/arrow/ipc/writer.cc [hunk 1]
```diff
@@ -222,8 +222,9 @@ class RecordBatchSerializer {
       RETURN_NOT_OK(
           result->Resize(actual_length + sizeof(int64_t), /* shrink_to_fit= */ true));
     }
-    *reinterpret_cast<int64_t*>(result->mutable_data()) =
-        bit_util::ToLittleEndian(prefixed_length);
+    int64_t prefixed_length_little_endian = bit_util::ToLittleEndian(prefixed_length);
```
Comment (cyb70289) [inline cpp/src/arrow/ipc/writer.cc:225]: you can use utils at https://github.com/apache/arrow/blob/main/cpp/src/arrow/util/ubsan.h#L81

Comment (mapleFU) [reply cpp/src/arrow/ipc/writer.cc:225]: done

File: cpp/src/arrow/ipc/writer.cc [hunk 2]
```diff
@@ -222,8 +222,9 @@ class RecordBatchSerializer {
...
+    util::SafeStore(result->mutable_data(), prefixed_length);
```
Comment (kou) [inline cpp/src/arrow/ipc/writer.cc:226]:
```suggestion
    util::SafeStore(result->mutable_data(), prefixed_length_little_endian);
```

Comment (mapleFU) [reply cpp/src/arrow/ipc/writer.cc]: Ahh, a stupid bug, nice catch, thanks!

**Review by kou — APPROVED** [review]
+1

**Comment by mapleFU** [comment]
Will merge in 2 days if no negative comments

**Pull request merged**